File: src/main/java/com/fretron/utils/PolylineUtils/OnlinePolylineEncoder.java
package com.fretron.utils.PolylineUtils;

/**
 * Created by sdhaker on 21/07/17.
 */

import com.fretron.Model.LitePosition;
import com.fretron.Model.PointAtTime;
import com.fretron.Model.TimeAwarePolyline;
import java.util.ArrayList;
import java.util.Iterator;
import java.util.List;

/**
 * This is a utility
 * For streaming we need a polyline line encoder to encode stream growing polyline
 *
 */
public class OnlinePolylineEncoder {
    private static  PolylineDecoder decoder = PolylineDecoder.getInstance();
    private static int floor1e5(double coordinate) {
        return (int) Math.floor(coordinate * 1e5);
    }

    private static String encodeSignedNumber(int num) {
        int sgn_num = num << 1;
        if (num < 0) {
            sgn_num = ~(sgn_num);
        }
        return (encodeNumber(sgn_num));
    }

    private static String encodeNumber(int num) {
        StringBuffer encodeString = new StringBuffer();
        while (num >= 0x20) {
            int nextValue = (0x20 | (num & 0x1f)) + 63;
            encodeString.append((char) (nextValue));
            num >>= 5;
        }
        num += 63;
        encodeString.append((char) (num));
        return encodeString.toString();
    }

    private static String encodeSignedNumber(long num) {
        long sgn_num = num << 1;
        if (num < 0) {
            sgn_num = ~(sgn_num);
        }
        return (encodeNumber(sgn_num));
    }

    private static String encodeNumber(long num) {
        StringBuilder encodeString = new StringBuilder();
        while (num >= 0x20) {
            long nextValue = (0x20 | (num & 0x1f)) + 63;
            encodeString.append((char) (nextValue));
            num >>= 5;
        }
        num += 63;
        encodeString.append((char) (num));
        return encodeString.toString();
    }



    public static String extendPolyline(String lineString , double lastLat , double lastLng , double newLat , double newLng){
        int lastlate5 = floor1e5(lastLat);
        int lastlnge5 = floor1e5(lastLng);
        int late5 = floor1e5(newLat);
        int lnge5 = floor1e5(newLng);
        int dlat = late5 - lastlate5;
        int dlng = lnge5 - lastlnge5;
        lineString = lineString +  (encodeSignedNumber(dlat) + encodeSignedNumber(dlng).toString());
        return lineString;
    }



    public static TimeAwarePolyline extendTimeAwarePolyline(TimeAwarePolyline polylineObj , double newLat , double newLng , long newLocationTime){
        if (polylineObj == null){ return  null;}
        PointAtTime lastPoint = polylineObj.getLastPoint();
        int lastlate5 = floor1e5(lastPoint.getLatitude());
        int lastlnge5 = floor1e5(lastPoint.getLongitude());
        int late5 = floor1e5(newLat);
        int lnge5 = floor1e5(newLng);
        int dlat = late5 - lastlate5;
        int dlng = lnge5 - lastlnge5;
        long dtime = newLocationTime - lastPoint.getTimestamp();

        if (dlat != 0 || dlng != 0){
            String extendedPolyline = (encodeSignedNumber(dlat).toString() + encodeSignedNumber(dlng).toString()) + encodeSignedNumber(dtime).toString() ;
            polylineObj.setPolyline(polylineObj.getPolyline() + extendedPolyline);
            lastPoint.setTimestamp(newLocationTime);
            lastPoint.setLatitude(newLat);
            lastPoint.setLongitude(newLng);
            polylineObj.setLastPoint(lastPoint);
            polylineObj.setTotalPoints(polylineObj.getTotalPoints()+1);
        }
        return polylineObj;
    }


    public static String extendPolyline(String lineString , LitePosition lastPos, LitePosition nextPos){
        int lastlate5 = floor1e5(lastPos.getLatitude());
        int lastlnge5 = floor1e5(lastPos.getLongitude());
        int late5 = floor1e5(nextPos.getLatitude());
        int lnge5 = floor1e5(nextPos.getLongitude());
        int dlat = late5 - lastlate5;
        int dlng = lnge5 - lastlnge5;
        lineString = lineString +  (encodeSignedNumber(dlat) + encodeSignedNumber(dlng).toString());
        return lineString;
    }




    public static TimeAwarePolyline extendTimeAwarePolyline(TimeAwarePolyline polylineObj ,LitePosition nextPos){
        if (polylineObj == null) return null;
        PointAtTime lastPoint = polylineObj.getLastPoint();
        int lastlate5 = floor1e5(lastPoint.getLatitude());
        int lastlnge5 = floor1e5(lastPoint.getLongitude());
        int late5 = floor1e5(nextPos.getLatitude());
        int lnge5 = floor1e5(nextPos.getLongitude());
        long lastTimestamp = lastPoint.getTimestamp();
        long timestamp = nextPos.getTime();
        int dlat = late5 - lastlate5;
        int dlng = lnge5 - lastlnge5;
        long dtime = timestamp - lastTimestamp;
        if (dlat != 0 || dlng != 0){
            String extendedPolyline = (encodeSignedNumber(dlat) + encodeSignedNumber(dlng).toString()) + encodeSignedNumber(dtime).toString() ;
            polylineObj.setPolyline(polylineObj.getPolyline() + extendedPolyline);
            lastPoint.setTimestamp(nextPos.getTime());
            lastPoint.setLatitude(nextPos.getLatitude());
            lastPoint.setLongitude(nextPos.getLongitude());
            polylineObj.setLastPoint(lastPoint);
            polylineObj.setTotalPoints(polylineObj.getTotalPoints()+1);
        }
        return polylineObj;
    }




    public static String mergePolylines(String polyline1 , String polyline2 , double lastlat1 , double lastlng1){
        double lastlat = lastlat1;
        double lastlng = lastlng1;
        List<Point> points  = decoder.decode(polyline2);
        for(Point point : points){
            polyline1 = extendPolyline(polyline1 , lastlat , lastlng , point.getLat() , point.getLng());
            lastlat =  point.getLat();
            lastlng =  point.getLng();
        }
        return polyline1;
    }



    public static TimeAwarePolyline mergeTimeAwarePolylines(TimeAwarePolyline polylineObj1 , TimeAwarePolyline polylineObj2){
        if (polylineObj1 != null && polylineObj2 != null){
            List<PointAtTime> points  = decoder.decodeTimeAwarePolyline(polylineObj2.getPolyline());
            for(PointAtTime point : points){
                polylineObj1 = extendTimeAwarePolyline(polylineObj1 ,point.getLatitude() ,point.getLongitude() ,point.getTimestamp());
            }
        }else if (polylineObj1 == null && polylineObj2 != null) return polylineObj2;

        return polylineObj1;
    }



    public static  String mergePolylines(String polyline1 ,String polyline2){
        if (polyline1.length() == 0) return polyline2;
        List<Point> points = decoder.decode(polyline1);
        Point lastPoint = points.get(points.size() - 1);
        double lastlat = lastPoint.getLat();
        double lastlng = lastPoint.getLng();
        points = decoder.decode(polyline2);
        for (Point point : points){
            polyline1 = extendPolyline(polyline1,lastlat,lastlng,point.getLat(),point.getLng());
            lastlat = point.getLat();
            lastlng = point.getLng();
        }
        return polyline1;
    }


    public static TimeAwarePolyline insertPointsIn(TimeAwarePolyline polylineObj ,List<Point> points ,long from ,long to){
        if (points.size() == 0 || polylineObj == null) return polylineObj;
        long timeDiff = to - from;
        long deltaTime = timeDiff / points.size();
        long nextLocationTime = from + deltaTime;
        for (Point point: points) {
            polylineObj = extendTimeAwarePolyline(polylineObj ,point.getLat() ,point.getLng() ,nextLocationTime);
            nextLocationTime += deltaTime;
        }
        return  polylineObj;
    }

    public static TimeAwarePolyline mergePolylineWithTimeAwarePolyline(TimeAwarePolyline polylineObj ,String polyline ,long newLocationTime){
        if (polyline.length() == 0 || polylineObj == null) return polylineObj;
        long lastPointTimestamp = polylineObj.getLastPoint().getTimestamp();
        long timeDiff = newLocationTime - lastPointTimestamp;
        List<Point> npoints = decoder.decode(polyline);
        if (npoints.size() > 0){
            long deltaTime = timeDiff / npoints.size();
            long nextLocationTime = lastPointTimestamp + deltaTime;
            for (Point point: npoints) {
                polylineObj = extendTimeAwarePolyline(polylineObj ,point.getLat() ,point.getLng() ,nextLocationTime);
                nextLocationTime += deltaTime;
            }
        }

        return  polylineObj;
    }


    public static String extendPolyline(Point[] list){
        String p = "";
        if (list.length > 0){
            double l = 0;
            double la = 0.0 ;
            for (Point aList : list) {
                p = extendPolyline(p, l, la, aList.getLat(), aList.getLng());
                l = aList.getLat();
                la = aList.getLng();
            }
        }
        return p;
    }


    public static TimeAwarePolyline extendTimeAwarePolyline(ArrayList<PointAtTime> list){
        TimeAwarePolyline polyline = null;
        if (list.size() > 0){
            polyline = new TimeAwarePolyline("","",new PointAtTime(0L,0d,0d),false,0);
            for (PointAtTime point : list) {
                polyline = extendTimeAwarePolyline(polyline, point.getLatitude(), point.getLongitude(), point.getTimestamp());
            }
        }
        return polyline;
    }


    public static TimeAwarePolyline extendTimeAwarePolyline(List<Point> list ,long startTime ,long endTime){
        TimeAwarePolyline polyline = null;
        if (list.size() > 0){
            long deltaTime = endTime - startTime / list.size();
            polyline = new TimeAwarePolyline("","",new PointAtTime(0L,0d,0d),false,0);
            for (Point point : list) {
                polyline = extendTimeAwarePolyline(polyline, point.getLat(), point.getLng(), startTime);
                startTime += deltaTime;
            }
        }
        return polyline;
    }


    public static TimeAwarePolyline extendTimeAwarePolylineWithPoints(TimeAwarePolyline polyline ,List<PointAtTime> list){
        if (list.size() > 0){
            for (PointAtTime point : list) {
                polyline = extendTimeAwarePolyline(polyline, point.getLatitude(), point.getLongitude(), point.getTimestamp());
            }
        }
        return polyline;
    }


    public static TimeAwarePolyline extendTimeAwarePolylineUsingPositions(List<LitePosition> list){
        TimeAwarePolyline polyline = null;
        if (list.size() > 0){
            polyline = new TimeAwarePolyline("","",new PointAtTime(0L,0d,0d),false,0);
            for (LitePosition point : list) {
                polyline = extendTimeAwarePolyline(polyline, point.getLatitude(), point.getLongitude(), point.getTime());
            }

        }
        return polyline;
    }



    public static String convertToGeoPolyline(TimeAwarePolyline timeAwarePolyline){
        String polyline = "";
        List<PointAtTime> list = decoder.decodeTimeAwarePolyline(timeAwarePolyline.getPolyline());
        Iterator<PointAtTime> it =  list.iterator();
        double lastLat = 0;
        double lastLng = 0;
        while (it.hasNext()){
            PointAtTime point = it.next();
            polyline = extendPolyline(polyline ,lastLat ,lastLng,point.getLatitude() ,point.getLongitude());
            lastLat = point.getLatitude();
            lastLng = point.getLongitude();
        }

        return polyline;
    }




}

File: src/main/java/com/fretron/Utils/Geospacial/PolylineUtils/OnlinePolylineEncoder.kt
package com.fretron.Utils.Geospacial.PolylineUtils

import com.fretron.Logger.Log
import com.fretron.Model.LitePosition
import com.fretron.Model.PointAtTime
import com.fretron.Model.TimeAwarePolyline
import java.util.*

object OnlinePolylineEncoder {

    init {
        Log.info("polyline encoder init called.")
    }

    private val decoder = PolylineDecoder.getInstance()
    private fun floor1e5(coordinate: Double): Int {
        return Math.floor(coordinate * 1e5).toInt()
    }

    private fun encodeSignedNumber(num: Int): String {
        var sNum = num shl 1
        if (num < 0) {
            sNum = sNum.inv()
        }
        return encodeNumber(sNum)
    }

    private fun encodeNumber(num: Int): String {
        var sNum = num
        val encodeString = StringBuffer()
        while (sNum >= 0x20) {
            val nextValue = (0x20 or (sNum and 0x1f)) + 63
            encodeString.append(nextValue.toChar())
            sNum = sNum shr 5
        }
        sNum += 63
        encodeString.append(sNum.toChar())
        return encodeString.toString()
    }

    private fun encodeSignedNumber(num: Long): String {
        var sgnNum = num shl 1
        if (num < 0) {
            sgnNum = sgnNum.inv()
        }
        return encodeNumber(sgnNum)
    }

    private fun encodeNumber(num: Long): String {
        var sNum = num
        val encodeString = StringBuilder()
        while (sNum >= 0x20) {
            val nextValue = (0x20.toLong() or (sNum and 0x1f)) + 63
            encodeString.append(nextValue.toChar())
            sNum = sNum shr 5
        }
        sNum += 63
        encodeString.append(sNum.toChar())
        return encodeString.toString()
    }


    fun extendPolyline(lineString: String, lastLat: Double, lastLng: Double, newLat: Double, newLng: Double): String {
        var polyline = lineString
        val lastLate5 = floor1e5(lastLat)
        val lastLnge5 = floor1e5(lastLng)
        val late5 = floor1e5(newLat)
        val lnge5 = floor1e5(newLng)
        val dlat = late5 - lastLate5
        val dlng = lnge5 - lastLnge5
        polyline += (encodeSignedNumber(dlat) + encodeSignedNumber(dlng))
        return polyline
    }


    fun extendTimeAwarePolyline(polylineObj: TimeAwarePolyline?,
                                newLat: Double,
                                newLng: Double,
                                newLocationTime: Long): TimeAwarePolyline? {
        if (polylineObj == null) {
            return null
        }
        val lastPoint = polylineObj.getLastPoint()
        val lastLat = floor1e5(lastPoint.getLatitude()!!)
        val lastLng = floor1e5(lastPoint.getLongitude()!!)
        val late5 = floor1e5(newLat)
        val lnge5 = floor1e5(newLng)
        val dlat = late5 - lastLat
        val dlng = lnge5 - lastLng
        val dtime = newLocationTime - lastPoint.getTimestamp()!!

        if (dlat != 0 || dlng != 0) {
            val extendedPolyline = encodeSignedNumber(dlat) + encodeSignedNumber(dlng) + encodeSignedNumber(dtime)
            polylineObj.setPolyline(polylineObj.getPolyline() + extendedPolyline)
            lastPoint.setTimestamp(newLocationTime)
            lastPoint.setLatitude(newLat)
            lastPoint.setLongitude(newLng)
            polylineObj.setLastPoint(lastPoint)
            polylineObj.setTotalPoints(polylineObj.getTotalPoints() + 1)
        }
        return polylineObj
    }


    fun extendPolyline(lineString: String, lastPos: LitePosition, nextPos: LitePosition): String {
        var polyline = lineString
        val lastLat = floor1e5(lastPos.getLatitude())
        val lastLng = floor1e5(lastPos.getLongitude())
        val late5 = floor1e5(nextPos.getLatitude())
        val lnge5 = floor1e5(nextPos.getLongitude())
        val dlat = late5 - lastLat
        val dlng = lnge5 - lastLng
        polyline += (encodeSignedNumber(dlat) + encodeSignedNumber(dlng))
        return polyline
    }


    fun extendTimeAwarePolyline(polylineObj: TimeAwarePolyline?, nextPos: LitePosition): TimeAwarePolyline? {
        if (polylineObj == null) return null
        val lastPoint = polylineObj.getLastPoint()
        val lastLat = floor1e5(lastPoint.getLatitude()!!)
        val lastLng = floor1e5(lastPoint.getLongitude()!!)
        val late5 = floor1e5(nextPos.getLatitude()!!)
        val lnge5 = floor1e5(nextPos.getLongitude()!!)
        val lastTimestamp = lastPoint.getTimestamp()!!
        val timestamp = nextPos.getTime()!!
        val dlat = late5 - lastLat
        val dlng = lnge5 - lastLng
        val dtime = timestamp - lastTimestamp
        if (dlat != 0 || dlng != 0) {
            val extendedPolyline = encodeSignedNumber(dlat) + encodeSignedNumber(dlng) + encodeSignedNumber(dtime)
            polylineObj.setPolyline(polylineObj.getPolyline() + extendedPolyline)
            lastPoint.setTimestamp(nextPos.getTime())
            lastPoint.setLatitude(nextPos.getLatitude())
            lastPoint.setLongitude(nextPos.getLongitude())
            polylineObj.setLastPoint(lastPoint)
            polylineObj.setTotalPoints(polylineObj.getTotalPoints()!! + 1)
        }
        return polylineObj
    }


    fun mergePolylines(p1: String, polyline2: String, lastlat1: Double, lastlng1: Double): String {
        var polyline1 = p1
        var lastlat = lastlat1
        var lastlng = lastlng1
        val points = decoder.decode(polyline2)
        for (point in points) {
            polyline1 = extendPolyline(polyline1, lastlat, lastlng, point.lat, point.lng)
            lastlat = point.lat
            lastlng = point.lng
        }
        return polyline1
    }


    fun mergeTimeAwarePolylines(polylineObj1: TimeAwarePolyline?,
                                polylineObj2: TimeAwarePolyline?): TimeAwarePolyline? {
        var timeAwarePolyline = polylineObj1
        if (timeAwarePolyline != null && polylineObj2 != null) {
            val points = decoder.decodeTimeAwarePolyline(polylineObj2.getPolyline())
            for (point in points) {
                timeAwarePolyline = extendTimeAwarePolyline(timeAwarePolyline,
                    point.getLatitude()!!,
                    point.getLongitude()!!,
                    point.getTimestamp()!!)
            }
        } else if (timeAwarePolyline == null && polylineObj2 != null) return polylineObj2

        return timeAwarePolyline
    }


    fun mergePolylines(polyline1: String, polyline2: String): String {
        var polyline = polyline1
        if (polyline.isEmpty()) return polyline2
        var points = decoder.decode(polyline)
        val lastPoint = points[points.size - 1]
        var lastlat = lastPoint.lat
        var lastlng = lastPoint.lng
        points = decoder.decode(polyline2)
        for (point in points) {
            polyline = extendPolyline(polyline, lastlat, lastlng, point.lat, point.lng)
            lastlat = point.lat
            lastlng = point.lng
        }
        return polyline
    }


    fun insertPointsIn(polylineObj: TimeAwarePolyline?, points: List<Point>, from: Long, to: Long): TimeAwarePolyline? {
        var timeAwarePolyline = polylineObj
        if (points.isEmpty() || timeAwarePolyline == null) return timeAwarePolyline
        val timeDiff = to - from
        val deltaTime = timeDiff / points.size
        var nextLocationTime = from + deltaTime
        for (point in points) {
            timeAwarePolyline = extendTimeAwarePolyline(timeAwarePolyline, point.lat, point.lng, nextLocationTime)
            nextLocationTime += deltaTime
        }
        return timeAwarePolyline
    }

    fun mergePolylineWithTimeAwarePolyline(polylineObj: TimeAwarePolyline?,
                                           polyline: String,
                                           newLocationTime: Long): TimeAwarePolyline? {
        var timeAwarePolyline = polylineObj
        if (polyline.isEmpty() || timeAwarePolyline == null) return timeAwarePolyline
        val lastPointTimestamp = timeAwarePolyline.getLastPoint().getTimestamp()
        val timeDiff = newLocationTime - lastPointTimestamp
        val nPoints = decoder.decode(polyline)
        if (nPoints.size > 0) {
            val deltaTime = timeDiff / nPoints.size
            var nextLocationTime = lastPointTimestamp + deltaTime
            for (point in nPoints) {
                timeAwarePolyline = extendTimeAwarePolyline(timeAwarePolyline, point.lat, point.lng, nextLocationTime)
                nextLocationTime += deltaTime
            }
        }

        return timeAwarePolyline
    }


    fun extendPolyline(list: Array<Point>): String {
        var p = ""
        if (list.isNotEmpty()) {
            var l = 0.0
            var la = 0.0
            for (aList in list) {
                p = extendPolyline(p, l, la, aList.lat, aList.lng)
                l = aList.lat
                la = aList.lng
            }
        }
        return p
    }


    fun extendTimeAwarePolyline(list: ArrayList<PointAtTime>): TimeAwarePolyline? {
        var polyline: TimeAwarePolyline? = null
        if (list.size > 0) {
            polyline = TimeAwarePolyline("", "", PointAtTime(0L, 0.0, 0.0), false, 0)
            for (point in list) {
                polyline = extendTimeAwarePolyline(polyline,
                    point.getLatitude()!!,
                    point.getLongitude()!!,
                    point.getTimestamp()!!)
            }
        }
        return polyline
    }


    fun extendTimeAwarePolyline(list: List<Point>, startTime: Long, endTime: Long): TimeAwarePolyline? {
        var time = startTime
        var polyline: TimeAwarePolyline? = null
        if (list.isNotEmpty()) {
            val deltaTime = endTime - time / list.size
            polyline = TimeAwarePolyline("", "", PointAtTime(0L, 0.0, 0.0), false, 0)
            for (point in list) {
                polyline = extendTimeAwarePolyline(polyline, point.lat, point.lng, time)
                time += deltaTime
            }
        }
        return polyline
    }


    fun extendTimeAwarePolylineWithPoints(polyline: TimeAwarePolyline?, list: List<PointAtTime>): TimeAwarePolyline? {
        var timeAwarePolyline = polyline
        if (list.isNotEmpty()) {
            for (point in list) {
                timeAwarePolyline = extendTimeAwarePolyline(timeAwarePolyline,
                    point.getLatitude(),
                    point.getLongitude(),
                    point.getTimestamp())
            }
        }
        return timeAwarePolyline
    }




    fun extendTimeAwarePolylineUsingPositions(list: List<LitePosition>): TimeAwarePolyline? {
        var polyline: TimeAwarePolyline? = null
        if (list.isNotEmpty()) {
            polyline = TimeAwarePolyline("", "", PointAtTime(0L, 0.0, 0.0), false, 0)
            for (point in list) {
                polyline =
                    extendTimeAwarePolyline(polyline, point.getLatitude()!!, point.getLongitude()!!, point.getTime()!!)
            }
        }
        return polyline
    }

    fun extendTimeAwarePolylineUsingPositions(polyline : TimeAwarePolyline?,list: List<LitePosition>): TimeAwarePolyline? {
        var timeAwarePolyline: TimeAwarePolyline? = polyline ?: TimeAwarePolyline("", "", PointAtTime(0L, 0.0, 0.0), false, 0)
        if (list.isNotEmpty()) {
            for (point in list) {
                timeAwarePolyline =
                    extendTimeAwarePolyline(timeAwarePolyline, point.getLatitude(), point.getLongitude(), point.getTime())
            }
        }
        return timeAwarePolyline
    }


    fun convertToGeoPolyline(timeAwarePolyline: TimeAwarePolyline): String {
        var polyline = ""
        val list = decoder.decodeTimeAwarePolyline(timeAwarePolyline.getPolyline())
        val it = list.iterator()
        var lastLat = 0.0
        var lastLng = 0.0
        while (it.hasNext()) {
            val point = it.next()
            polyline = extendPolyline(polyline, lastLat, lastLng, point.getLatitude()!!, point.getLongitude()!!)
            lastLat = point.getLatitude()!!
            lastLng = point.getLongitude()!!
        }

        return polyline
    }


}
